### DISCUSSION TITLE
GOPLS Organize Imports

### DISCUSSION THREAD

**Discussion by samimb** [joaotavora/eglot#1117] [Q&A] [answered]
I’m trying to get the code action of organize imports to work with before-save-hook. Specifically for `go-mode`.
I’ve been searching around and found a couple of issues later linking to a solution similar to this:

```
(defun sami/eglot-organize-imports ()
  (call-interactively 'eglot-code-action-organize-imports))
(defun sami/before-saving-go ()
  (add-hook 'before-save-hook #'eglot-format-buffer -10 t)
  (add-hook 'before-save-hook #'sami/eglot-organize-imports nil t))
(add-hook 'go-mode-hook #'sami/before-saving-go)
```

Problem seems to be, running the function `eglot-code-actions` and selecting `Organize imports` works fine, but running the stand-alone function of `eglot-code-action-organize-imports` does not yield anything.

Test case:
Open emacs with `emacs -q` for an empty config
```elisp
(require 'go-mode)
(require 'eglot)
(add-hook 'go-mode-hook #'eglot-ensure)
```
In any go file, add something like `fmt.Println(“hello”)` should import the package `fmt` automatically, deleting or commenting the line does not delete the imported package `fmt`.   `eglot-code-action-organize-imports` does not delete the imported package `fmt`. Running the function `eglot-code-actions` and selecting `Organize imports` DOES delete the package `fmt` from imports.

Here is the `client-request` and `server-response`.

```elisp
[client-request] (id:613) Mon Dec  5 23:43:48 2022:
(:jsonrpc "2.0" :id 613 :method "textDocument/codeAction" :params
	  (:textDocument
	   (:uri "file:///Users/sami/Code/Go/src/sami/something/temp.go")
	   :range
	   (:start
	    (:line 5 :character 6)
	    :end
	    (:line 5 :character 10))
	   :context
	   (:diagnostics
	    []
	    :only
	    ["source.organizeImports"])))
[server-reply] (id:613) Mon Dec  5 23:43:48 2022:
(:jsonrpc "2.0" :result
	  [(:title "Organize Imports" :kind "source.organizeImports" :edit
		   (:documentChanges
		    [(:textDocument
		      (:version 172 :uri "file:///Users/sami/Code/Go/src/sami/something/temp.go")
		      :edits
		      [(:range
			(:start
			 (:line 3 :character 2)
			 :end
			 (:line 3 :character 2))
			:newText "bytes\"\n	\"crypto/tls\"\n	\"encoding/json\"\n	\"")
		       (:range
			(:start
			 (:line 4 :character 2)
			 :end
			 (:line 4 :character 2))
			:newText "go/token\"\n	\"")
		       (:range
			(:start
			 (:line 5 :character 10)
			 :end
			 (:line 5 :character 10))
			:newText "\"\n	\"sami/something/models\"\n	\"sami/something/tool")])]))]
	  :id 613)
```

Single difference between the usage of `eglot-code-action-organize-imports` and not is 
```elisp
...
            :only
            ["source.organizeImports"])))
...
```

**Comment by samimb** [comment]
emacs-28.2 (emacs-mac, railwaycat/emacsmacport/emacs-mac)
eglot-20221020.1010 (melpa)

**Comment by noyv** [comment]
The problem maybe intronduced by commit https://github.com/joaotavora/eglot/commit/2599cd0d7951c6436b01e950110c98ee6375786a 

The flow is like this:  eglot-code-action-organize-imports -> macro eglot--code-action(but no interactive) -> eglot-code-actions -> skip eglot--read-execute-code-action

This is the reason why the command has no effect. My elisp is poor,  then the guess may be wrong.  the simplest solution for me is to comment the line https://github.com/joaotavora/eglot/blob/master/eglot.el#L3198 as I only use the code action for importing.

**Comment by joaotavora** [comment, marked as answer]
This bug has been fixed in Emacs 29: https://git.savannah.gnu.org/cgit/emacs.git/commit/?h=emacs-29&id=9d3fdf7e0d4c9355b2466aca51938291ea4e26c7
